### PR TITLE
Add tests and benchmarks

### PR DESCRIPTION
**Pull request by Jake-Shadle**
Resolves: #15

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -3,3 +3,6 @@ members = [
     "lib",
     "cli",
 ]
+
+[patch.crates-io]
+img_hash = { git = "https://github.com/EmbarkStudios/img_hash.git", rev = "c40da78" }
```
Comment (repi) [inline Cargo.toml]: Add a link to the PR here? https://github.com/abonander/img_hash/pull/30

File: lib/src/lib.rs [hunk 2]
```diff
@@ -65,6 +65,7 @@ struct Parameters {
     output_size: (u32, u32),
     guide_alpha: f32,
     random_resolve: Option<u64>,
+    max_thread_count: Option<u32>,
```
Comment (repi) [inline lib/src/lib.rs]: Minor nitpick: technically as the # threads is a count it probably should be `usize` type. Though I doubt one would use more than 4 million threads, but you don't have to cast then in the map on usage.

```suggestion
    max_thread_count: Option<usize>,
```

File: lib/tests/diff.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                let distance = expected_hash.dist_ratio(&gen_hash);
+                let txt_gen = gen_hash.to_base64();
+
+                assert_eq!($expected, txt_gen, "images hashes differed by {}", distance);
+            }
+        }
+    };
+}
+
+diff_runs!(single_example, "JKc2MqWo1iNWeJ856Ty6+a1M", {
```
Comment (repi) [inline lib/tests/diff.rs]: Should add some docs in the README and link here how to re-generate the hash and that one needs to update it when doing changes to the algorithm

Comment (Jake-Shadle) [reply lib/tests/diff.rs]: I put it into the diff.rs file itself as the README is more of a top level user guide, only people who actually make modifications would ever need to know about the tests.

Comment (repi) [reply lib/tests/diff.rs]: ok, yeah that should work well

**Pull request merged**